docs/src/tutorials/getting_started/sum_if.jl

```diff
@@ -0,0 +1,245 @@
+# mathematical formulation of a network flow problem.
+
+# The downside to this formulation is subtle. Behind the scenes, the JuMP
+# `@constraint` macro expands to something like:
+
+model = Model()
+@variable(model, flows[e in edges] >= 0)
+for n in nodes
+    flow_in = AffExpr(0.0)
+    for (i, j) in edges
+        if j == n
+            add_to_expression!(flow_in, flows[(i, j)])
+        end
+    end
+    flow_out = AffExpr(0.0)
+    for (i, j) in edges
+        if i == n
+            add_to_expression!(flow_out, flows[(i, j)])
+        end
+    end
+    @constraint(model, flow_in - flow_out == demand(n))
+end
+
+# This formulation includes two for-loops, with a loop over every edge (twice) for
+# every node. The [big-O notation](https://en.wikipedia.org/wiki/Big_O_notation)
+# of the runtime is ``O(|nodes| \times |edges|)``. If
+# you have a large number of nodes and a large number of edges, the runtime of
+# this loop can be large.
+
+# Let's build a function to benchmark our formulation:
+
+function build_naive_model(nodes, edges, demand)
+    model = Model()
+    @variable(model, flows[e in edges] >= 0)
+    @constraint(
+        model,
+        [n in nodes],
+        sum(flows[(i, j)] for (i, j) in edges if j == n) -
+        sum(flows[(i, j)] for (i, j) in edges if i == n) == demand(n)
+    )
+    return model
+end
+
+nodes, edges, demand = build_random_graph(1_000, 2_000)
+@elapsed build_naive_model(nodes, edges, demand)
+
+# A good way to benchmark is to measure the runtime across a wide range of input
+# sizes. From our big-O analysis, we should expect that doubling the number of
+# nodes and edges results in a 4x increase in the runtime.
+
+run_times = Float64[]
+factors = 1:10
+for factor in factors
+    graph = build_random_graph(1_000 * factor, 5_000 * factor)
+    push!(run_times, @elapsed build_naive_model(graph...))
+end
+Plots.plot(; xlabel = "Factor", ylabel = "Runtime [s]")
+Plots.plot!(factors, run_times; label = "Actual")
+a, b = hcat(ones(10), factors .^ 2) \ run_times
+Plots.plot!(factors, a .+ b * factors .^ 2; label = "Quadratic fit")
+
+# As expected, the runtimes demonstrate quadratic scaling: if we double the
+# number of nodes and edges, the runtime increases by a factor of four.
+
+# ## Caching
+
+# We can improve our formulation by caching the list of incoming and outgoing
+# nodes for each node `n`:
+
+out_nodes = Dict(n => Int[] for n in nodes)
+in_nodes = Dict(n => Int[] for n in nodes)
+for (i, j) in edges
+    push!(out_nodes[i], j)
+    push!(in_nodes[j], i)
+end
+
+# with the corresponding change to our model:
+
+model = Model()
+@variable(model, flows[e in edges] >= 0)
+@constraint(
+    model,
+    [n in nodes],
+    sum(flows[(i, n)] for i in in_nodes[n]) -
+    sum(flows[(n, j)] for j in out_nodes[n]) == demand(n)
+);
+
+# The benefit of this formulation is that we now loop over `out_nodes[n]`
+# rather than `edges` for each node `n`. If the graph is sparse, so that the
```

Comment:
You can say that now, the complexity is `O(|edges|)` instead of `O(|nodes| \times |edges|)` because you will see each edge only twice. So you have a big win even if the graph is not sparse if I'm not mistaken

Reply:
If it's sparse, i.e., number of edges is a linear function in terms of num of nodes you go from quadratic to linear. If its not sparse, you go from cubic to quadratic.